src/Microsoft.Azure.WebJobs.Host/Bindings/OpenType.cs

```diff
@@ -6,28 +6,328 @@
 
 namespace Microsoft.Azure.WebJobs.Host.Bindings
 {
+    // Ensures that open Types match consistently 
+    public class OpenTypeMatchContext
+    {
+        internal Type _match;
+
+        internal bool CheckArg(Type type)
+        {
+            if (_match == null)
+            {
+                _match = type;
+                return true;
+            }
+            if (_match == type)
+            {
+                return true;
+            }
+            return false;
+        }
+    }
+
     /// <summary>
-    /// Placeholder to use with converter manager for describing generic types.     
+    /// Placeholder to use with converter manager for describing generic types.
     /// Derived classes can override IsMatch to provide a constraint. 
     ///  OpenType matches any type. 
     ///  MyDerivedType matches any type where IsMatch(type) is true. 
     /// Also applies to generics such as: 
     ///  GenericClass&lt;OpenType&gt; 
     /// </summary>
-    [Obsolete("Not ready for public consumption.")]
     [DebuggerDisplay("{GetDisplayName()}")]
     public abstract class OpenType
     {
         /// <summary>
+        /// Return true if and only if given type matches.
+        /// If the OpenType represents an exact type, then this is just Type.Equals.
+        /// </summary>
+        /// <param name="type">Type to check</param>
+        /// <returns></returns>
+        public bool IsMatch(Type type)
+        {
+            var context = new OpenTypeMatchContext();
+            return this.IsMatch(type, context);
+        }
+
+        /// <summary>
         /// Return true if and only if given type matches. 
+        /// This is constrained by the current matching context. 
+        /// If the OpenType represents an exact type, then this is just Type.Equals.
         /// </summary>
         /// <param name="type">Type to check</param>
         /// <returns></returns>
-        public abstract bool IsMatch(Type type);
+        public abstract bool IsMatch(Type type, OpenTypeMatchContext context);
```

Comment:
>OpenTypeMatchContext [](start = 48, length = 20)

Requiring this param is a breaking change. But very few extensions actually derive from OT directly.

Reply:
Should this be a protected abstract, since it is just the implementation for IsMatch(Type)? Do they both need to be public?